gax/src/main/java/com/google/api/gax/batching/FlowController.java

```diff
@@ -213,25 +234,106 @@ public void release(long elements, long bytes) {
       outstandingElementCount.release(elements);
     }
     if (outstandingByteCount != null) {
-      // Need to return at most as much bytes as it can be drawn.
-      long permitsToReturn = Math.min(bytes, maxOutstandingRequestBytes);
-      outstandingByteCount.release(permitsToReturn);
+      outstandingByteCount.release(bytes);
+    }
+  }
+
+  /**
+   * Increase flow control limits to allow extra elementSteps elements and byteSteps request bytes
+   * before enforcing flow control.
+   */
+  @InternalApi("For google-cloud-java client use only")
+  public void increaseThresholds(long elementSteps, long byteSteps) {
+    Preconditions.checkArgument(elementSteps >= 0);
+    Preconditions.checkArgument(byteSteps >= 0);
+    synchronized (updateLimitLock) {
+      if (outstandingElementCount != null) {
+        long actualStep = Math.min(elementSteps, maxElementCountLimit - currentElementCountLimit);
+        currentElementCountLimit += actualStep;
```

Comment:
This value(s) is updated under the lock, which means they must also be accessed under the lock in other places (at least in the `getCurrentElementCountLimit()` method, otherwise the lock does not serve its purpose. 

I.e. it is not enough so synchronize all the writes to a value, we also must all the synchronize reads from the value, otherwise, since reads are not synchronized, they can be done right when write is happening defeatnign the whole purpose of synchronization. 

The same comment applies to `currentRequestBytesLimit` value and probably to  the `outstandingByteCount` as a whole (because a mutative method .release() is called on outstandingByteCount under lock).